alicloud/resource_alicloud_cs_kubernetes_autoscaler.go

```diff
@@ -0,0 +1,552 @@
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-alicloud/alicloud/connectivity"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	clusterAutoscaler          = "cluster-autoscaler"
+	defaultAutoscalerNamespace = "kube-system"
+	defaultScalingGroupTag     = "k8s.aliyun.com"
+	defaultServiceAccountName  = "admin"
+	defaultAutoscalerImage     = "registry-vpc.%s.aliyuncs.com/acs/autoscaler:v1.3.1-7369cf1"
+	LabelPattern               = "k8s.io/cluster-autoscaler/node-template/label/"
+	TaintPattern               = "k8s.io/cluster-autoscaler/node-template/taint/"
+)
+
+// nodePool defines the struct of scaling group params
+type nodePool map[string]string
+
+// userKubeConf defines the struct of response of api
+type userKubeConf struct {
+	Config string `json:"config"`
+}
+
+// resourceAlicloudCSKubernetesAutoscaler defines the schema of resource
+func resourceAlicloudCSKubernetesAutoscaler() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAlicloudCSKubernetesAutoscalerCreate,
+		Read:   resourceAlicloudCSKubernetesAutoscalerRead,
+		Update: resourceAlicloudCSKubernetesAutoscalerUpdate,
+		Delete: resourceAlicloudCSKubernetesAutoscalerDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(90 * time.Minute),
+			Update: schema.DefaultTimeout(60 * time.Minute),
+			Delete: schema.DefaultTimeout(60 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"nodepools": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"taints": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"labels": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+				MaxItems: 30,
+			},
+			"utilization": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"cool_down_duration": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"defer_scale_in_duration": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+// resourceAlicloudCSKubernetesAutoscalerCreate define how to create autoscaler
+func resourceAlicloudCSKubernetesAutoscalerCreate(d *schema.ResourceData, meta interface{}) error {
+
+	var clusterId string
+	if value, ok := d.GetOk("cluster_id"); ok == true {
+		clusterId = value.(string)
+	} else {
+		return WrapError(fmt.Errorf("clusterId must be provided instead of %s", "nil"))
+	}
+
```

Comment:
The line 105-109 can be instead by setting `cluster_id` to `required.`

Reply:
Because the `cluster_id` is required, the line 104-109 can be instead by `clusterId := d.Get("cluster_id").(string)`